Add net.lutris.Lutris

RobLoach opened the pull request:
[Lutris](https://lutris.net/) is an Open Source gaming platform for Linux. It installs and launches games so you can start playing without the hassle of setting up your games.

This pull request is the result of @gasinvein's brilliant git-fu to combine the git history from two different sources:

- [Gasinvein's Pull Request](https://github.com/flathub/flathub/pull/919)
- [net.lutris.Lutris repository](https://github.com/RobLoach/net.lutris.Lutris)

AsciiWolf commented:
bot, build net.lutris.Lutris

gasinvein commented:
@RobLoach please enable issues for the new repository, and transfer (is it possible for non-organisation?) there issues from the old one.
Also I'd suggest to make the new repository Lutris-specific by renaming it and changing default branch, but it's up to you.

LeandroStanger commented:
**Add**
- [ ] psmisc (or the package providing 'fuser')
- [ ] pz7zip (or the package providing '7z')
- [ ] curl
- [ ] fluid-soundfont-gs (or other soundfonts for MIDI support)
- [ ] cabextract (if needed, to install Windows games)
- [ ] x11-xserver-utils (or the package providing 'xrandr', if you are running Xorg, if you are not, you will depend on the GnomeDesktop bindings to fetch screen resolutions on Wayland, the GnomeDesktop library is not directly related to the Gnome desktop and is only used as a xrandr replacement.)
- [ ] libc6-i386 and lib32gcc1 for 32bit games support
- [ ] The 32bit OpenGL driver for your graphics card

RobLoach commented:
- Updated [the issues](https://github.com/RobLoach/net.lutris.Lutris/issues)
- Renamed [the old repo as net.lutris.Lutris.old](https://github.com/RobLoach/net.lutris.Lutris.old)
- Renamed [the repo to net.lutris.Lutris](https://github.com/RobLoach/net.lutris.Lutris/issues)

gasinvein commented:
Added a [milestone](/RobLoach/net.lutris.Lutris/milestone/1) to track blocking issues.

gasinvein commented:
bot, build net.lutris.Lutris

gasinvein commented:
Please note that 32-bit compatibility is currently broken due to recent glibc update in the FD.o runtime. Either downgrade the `org.freedesktop.Platform.Compat.i386` extension or wait for Gnome runtime update.

LeandroStanger commented:
`FileNotFoundError: [Errno 2] No such file or directory: '/sysroot/home/leandro/.var/app/net.lutris.Lutris/data/lutris/runners/wine/tkg-4.4-x86_64/bin/wine': '/sysroot/home/leandro/.var/app/net.lutris.Lutris/data/lutris/runners/wine/tkg-4.4-x86_64/bin/wine'
`

gasinvein commented:
@LeandroStanger yes, 32-bit wine wouldn't work currently, see https://github.com/flathub/flathub/pull/926#issuecomment-478231140

LeandroStanger commented:
Dark theme works now
![Captura de tela de 2019-03-30 08-00-15](https://user-images.githubusercontent.com/26097419/55275243-f68b4980-52c1-11e9-9fc1-961100a92c88.png)

gasinvein commented:
@LeandroStanger It's good, yet we didn't do anything about it.

AsciiWolf commented:
bot, build net.lutris.Lutris

AsciiWolf commented:
`2019-03-31 14:55:46,383: glxinfo is not available on your system, unable to detect driver version`

Glxinfo seems to be missing.

And when trying to install any 32-bit application, I am getting the same _No such file or directory_  error [as LeandroStanger](https://github.com/flathub/flathub/pull/926#issuecomment-478231960). So we have to wait for the GNOME runtime to be updated.

gasinvein commented:
@AsciiWolf Please open an issue about missing glxinfo (same for any problem you would find).

gasinvein commented:
bot, build net.lutris.Lutris

valentindavid commented:
I see this uses Gnome runtime and Freedesktop SDK 32 bits extensions at the same time. Gnome runtime releases do not necessarily follow the Freedesktop SDK. I am afraid that this will be a source of issues. For example incompatible configuration files. But also possibly an incompatible symlink to the dynamic loader. I would recommend either:
 - Get Gnome to provide a Compat.i386 extension.
 - Build what you need from Gnome runtime directly into the application.

gasinvein commented:
@valentindavid Yeah I alredy have a branch for building Lutris with FD.o runtime. The problem is that Luris requires WebKitGTK, which requires extreme amount of time to build.

LeandroStanger commented:
@RobLoach Just add 2 runtime?

LeandroStanger commented:
@gasinvein Just add 2 runtime?

jatin-cbs commented:
this error on every game install on lutris flatpak
![Screenshot from 2019-04-01 17-48-24](https://user-images.githubusercontent.com/40650341/55326883-74a84580-54a6-11e9-8bff-47694c587238.png)

gasinvein commented:
@jatin-cbs opened RobLoach/net.lutris.Lutris#20

jatin-cbs commented:
This error whenever we click play

gasinvein commented:
@jatin-cbs please post additional comments at RobLoach/net.lutris.Lutris#20, this PR is not a suitable place to discuss specific issues.

RobLoach commented:
@LeandroStanger Why remove README.md? I've found it handy to reference as developer documentation when building locally.

LeandroStanger commented:
@RobLoach I thought it was no longer needed

gasinvein commented:
Well, the `README.md` was hamless, yet slightly outdated.

gasinvein commented:
Is there any way to use the buildbot to build another branch, which is not source of this PR?
It would be very helpful if we could use the buildbot to build RobLoach/net.lutris.Lutris#21

nedrichards commented:
@gasinvein not when you're in a PR, but you can easily build it locally and test there.

gasinvein commented:
@nedrichards The problem is that those branch includes WebKitGTK, which requires too much time to build locally on my PC.

LeandroStanger commented:
bot, build net.lutris.Lutris

LeandroStanger commented:
bot, build net.lutris.Lutris

AsciiWolf commented:
bot, build net.lutris.Lutris

gasinvein commented:
bot, build net.lutris.Lutris

Saroufim commented:
32-bit wine now runs well but the controller issue I reported earlier still persists. Dualshock 4 is not detected properly while the RPM-based Lutris detects it just fine (And flatpaked Steam too).

gasinvein commented:
@Saroufim can you please open an issue?

AsciiWolf commented:
I have tried few 32-bit Wine games and they all worked fine! :-)

However, none of the native games that I tried worked.

1. Minecraft could not find a renderer, but I suspect this being a Java issue:
```
Graphics Device initialization failed for :  es2, sw
Error initializing QuantumRenderer: no suitable pipeline found
java.lang.RuntimeException: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
	at com.sun.javafx.tk.quantum.QuantumRenderer.getInstance(QuantumRenderer.java:280)
	at com.sun.javafx.tk.quantum.QuantumToolkit.init(QuantumToolkit.java:221)
	at com.sun.javafx.tk.Toolkit.getToolkit(Toolkit.java:205)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:209)
	at javafx.embed.swing.JFXPanel.initFx(JFXPanel.java:215)
	at javafx.embed.swing.JFXPanel.<init>(JFXPanel.java:230)
	at net.minecraft.launcher.ui.tabs.website.JFXBrowser.<init>(JFXBrowser.java:31)
	at net.minecraft.launcher.ui.tabs.WebsiteTab.selectBrowser(WebsiteTab.java:45)
	at net.minecraft.launcher.ui.tabs.WebsiteTab.<init>(WebsiteTab.java:23)
	at net.minecraft.launcher.ui.tabs.LauncherTabPanel.<init>(LauncherTabPanel.java:18)
	at net.minecraft.launcher.ui.LauncherPanel.<init>(LauncherPanel.java:37)
	at net.minecraft.launcher.SwingUserInterface.initializeFrame(SwingUserInterface.java:121)
	at net.minecraft.launcher.Launcher.<init>(Launcher.java:114)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at net.minecraft.bootstrap.Bootstrap.startLauncher(Bootstrap.java:240)
	at net.minecraft.bootstrap.Bootstrap.execute(Bootstrap.java:124)
	at net.minecraft.bootstrap.Bootstrap.main(Bootstrap.java:381)
Caused by: java.lang.RuntimeException: Error initializing QuantumRenderer: no suitable pipeline found
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.init(QuantumRenderer.java:94)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:124)
	at java.lang.Thread.run(Thread.java:745)
FATAL ERROR: net.minecraft.bootstrap.FatalBootstrapError: Unable to start: java.lang.reflect.InvocationTargetException
	at net.minecraft.bootstrap.Bootstrap.startLauncher(Bootstrap.java:242)
	at net.minecraft.bootstrap.Bootstrap.execute(Bootstrap.java:124)
	at net.minecraft.bootstrap.Bootstrap.main(Bootstrap.java:381)
```

2. SuperTuxKart could not find _libbluetooth.so.3_:
`bin/supertuxkart: error while loading shared libraries: libbluetooth.so.3: cannot open shared object file: No such file or directory`

3. OpenArena could not find _libSDL-1.2.so.0_:
`/home/asciiwolf/Games/openarena/openarena.x86_64: error while loading shared libraries: libSDL-1.2.so.0: cannot open shared object file: No such file or directory`

P.S. I have found a small issue with the Lutris desktop launcher. The "Lutris" launch entry is displayed two times for some reason:
![Lutris menu item](https://user-images.githubusercontent.com/2339930/55487755-fe831a80-562e-11e9-8ca4-568e13c72b94.png)

AsciiWolf commented:
The missing libSDL-1.2 will probably be an issue for more games, but could be easily fixed by adding these submodules to the Lutris Flatpak manifest:
```
- shared-modules/glu/glu-9.0.0.json
- shared-modules/SDL/SDL-1.2.15.json
```
(Optionally also _SDL_mixer_, _SDL_image_, _SDL_net_, _SDL_pango_, _SDL_ttf_ etc.)

The SuperTuxKart and Minecraft issues are probably safe to ignore for now. SuperTuxKart is available on Flathub and Minecraft doesn't work properly in a Flatpak anyway.

AsciiWolf commented on net.lutris.Lutris.yml:
```diff
@@ -0,0 +1,366 @@
+      - /lib/mesa-demos
+    modules:
+      - shared-modules/glew/glew.json
+      - shared-modules/glu/glu-9.0.0.json
```
```suggestion
      - shared-modules/glu/glu-9.0.0.json
      - shared-modules/SDL/SDL-1.2.15.json
```

gasinvein replied:
Here shouldn't be extra indentation level, glu is a dependency of mesa-demos, while sdl is not.
And, better, [open a PR](/RobLoach/net.lutris.Lutris/pulls).

AsciiWolf replied:
Oops, you're right.

AsciiWolf replied:
Anyway, I think it's fine to leave this unfixed for now. Most of the native games available via Lutris are also available on Flathub and and Lutris is primarily used for Wine/emulated games, not the native ones.

AsciiWolf replied:
Also, the best way to fix this would, in my opinion, be adding the missing libs to the [Lutris runtime](https://github.com/lutris/lutris/wiki/Lutris-Runtime) instead of including them in a Flatpak.

AsciiWolf commented:
LGTM!

LeandroStanger commented:
bot, build net.lutris.Lutris

TomasTomecek commented:
I just tried this and gotta say it works pretty well: installed wine steam and it runs nicely. Tried installing Path of exile game and it's done. Sadly I can't run it because it requires libgcrypt:
```
0153:err:module:load_builtin_dll failed to load .so lib for builtin L"WLDAP32.dll": libgcrypt.so.11: cannot open shared object file: No such file or directory
```

Just looked at the comments above and @AsciiWolf reported something similar: does this mean that the library needs to be included in the flatpak manifest or should lutris handle this?

gasinvein commented:
bot, build net.lutris.Lutris

Saroufim commented:
Either the Lutris runtime has to handle it or a gaming runtime has to be assembled for Flatpak (one that ideally includes all the possible libraries needed by wine too). Bundling all those libraries into the Lutris flatpak is an option too but I would say it's less than ideal.

gasinvein commented:
@TomasTomecek @Saroufim I've opened RobLoach/net.lutris.Lutris#15 to track missing wine dependencies.
The main problem is that flatpak bundles are single-arch. Thus, we can only add libraries for the same arch the app is built for - in most cases, it's x86_64. So this would help only for pure 64-bit games.

eszlari commented:
> The main problem is that flatpak bundles are single-arch. Thus, we can only add libraries for the same arch the app is built for - in most cases, it's x86_64. So this would help only for pure 64-bit games.

Relevant bug report: https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/348

LeandroStanger commented:
bot, build net.lutris.Lutris

nanonyme commented on net.lutris.Lutris.yml:
```diff
@@ -0,0 +1,364 @@
+  - --filesystem=~/Games:create
+  - --persist=~/.wine
+  - --filesystem=xdg-data/Steam:ro
+  - --filesystem=~/.var/app/com.valvesoftware.Steam:ro
```
Has this been tested? I've noted Flatpak doesn't allow apps to see each other's sandboxes even with --filesystem=home.

gasinvein replied:
Yes, it's working as intended. Gnome Games does [similar thing](https://github.com/flathub/org.gnome.Games/blob/1be076089e333bccc2d767be6d352e7551a767f3/org.gnome.Games.json#L37).

gasinvein commented:
bot, build net.lutris.Lutris

barthalion commented:
bot, build net.lutris.Lutris

barthalion commented:
LGTM. Is someone from upstream interested in getting involved?

gasinvein commented:
@barthalion Upstream is informed lutris/lutris#701, but I'm sure if interested. Better ask @RobLoach, AFAIK he is/was in contact with upstream.

gasinvein commented:
Is it possible to transfer the existing repository to flathub organization instead of creating new one? It would be really nice to keep all current issues/PRs.

nedrichards commented:
@gasinvein it doesn't appear to be easy, and wouldn't be our preferred way of doing things.

RobLoach commented:
@TingPing owns moving the repo to the flathub organization. There are still a [few issues present in the backlog](https://github.com/RobLoach/net.lutris.Lutris/issues), do we need to address those prior? I'm good with skipping them for now :+1:

nedrichards commented:
@RobLoach well, the ones with the initial release tag looks like they should be blockers. Or otherwise, have that tag removed.

gasinvein commented:
@nedrichards I wasn't able to reproduce RobLoach/net.lutris.Lutris#20, but I guess it's fixed after switching to FD.o runtime.
And RobLoach/net.lutris.Lutris#15 can't be truly fixed until we have [cross-compiling Sdk](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/348), as wine needs depends on libraries of both architectures.

LeandroStanger commented:
bot, build net.lutris.Lutris

LeandroStanger commented:
bot, build net.lutris.Lutris

LeandroStanger commented:
bot, build net.lutris.Lutris

LeandroStanger commented:
bot, build net.lutris.Lutris

LeandroStanger commented:
bot, build net.lutris.Lutris

gasinvein commented:
bot, build net.lutris.Lutris

gasinvein commented:
So, @RobLoach @LeandroStanger, what do you think, is this ready to publish?

valentindavid commented:
For me it complains about: 
```
2019-05-03 18:24:28,651: i386 libGL.so.1 missing (needed by opengl)
2019-05-03 18:24:28,651: i386 libvulkan.so.1 missing (needed by vulkan)
```

But `/usr/lib/i386-linux-gnu/libvulkan.so.1` and `/usr/lib/i386-linux-gnu/libGL.so.1` are present.

gasinvein commented:
@valentindavid Yeah, `ldconfig -p` inside flatpak sandbox doesn't list anything from the `.Compat.i386` extension for some reason. But libraries from there are still loaded as usual, so these errors are safe to ignore.

valentindavid commented:
You need to add it in `/app/etc/ld.so.conf`

valentindavid commented:
See with Steam: https://github.com/flathub/com.valvesoftware.Steam/blob/master/ld.so.conf

gasinvein commented:
> You need to add it in `/app/etc/ld.so.conf`

Just to make those warnings go away? For normal work of 32-bit apps this is not needed since FD.o runtime 18.08.

valentindavid commented:
Yes, it is needed to speed up loading. Please add it.

valentindavid commented:
And also, there is a popup window that also complains about it. I think it is quite annoying for the users.

gasinvein commented:
bot, build net.lutris.Lutris

valentindavid commented:
Thank you @gasinvein. It works now.

Saroufim commented:
Controllers and libretro runners still don't work. This will impact how people perceive this if it came with missing features upon release.

gasinvein commented:
@Saroufim Many prebuilt runners (including retroarch) probably won't work well. I hope Lutris devs would consider building them against FD.o Sdk to ensure ABI compatibility. Or we would build the runner as flatpak extensions for the app.

valentindavid commented:
The only missing thing for retroarch is `libudev.so.1`. You can add it from the shared modules.

valentindavid commented:
It needs also `libv4l2.so.0`. But we could probably build that as well on the app.

valentindavid commented:
Ah right. It is `libudev.so.0` provided by the shared modules. Strange that they still link to libudev and need a newer version. Usually it is not really used. It just needs to link to it.

gasinvein commented:
@valentindavid See RobLoach/net.lutris.Lutris#40

gasinvein commented:
bot, build net.lutris.Lutris

gasinvein commented:
bot, build net.lutris.Lutris

gasinvein commented:
bot, build net.lutris.Lutris

LeandroStanger commented:
```
So, @RobLoach @LeandroStanger, what do you think, is this ready to publish?
```
Publish

RobLoach commented:
There's still one issue marked in the "initial-release" milestone: [Add Wine dependencies](https://github.com/RobLoach/net.lutris.Lutris/issues/15).

gasinvein commented:
@RobLoach As I noted previously, we can't currently fix it anyway, so removing from the milestone.

gasinvein commented:
bot, build net.lutris.Lutris

gasinvein commented:
Can we publish on beta branch for now, and move to stable branch once it's 100% ready? This would be good for us, testers and other enthusiasts, but wouldn't confuse normal users.
Currently, there are some changes in upstream projects we are waiting for:
- The Lutris patches are waiting for review by Lutris devs (lutris/lutris#2106)
- Cross-compilling SDK is needed to build multiarch wine (https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/348)
- 32-bit compatibility extension for Gnome runtime would ease things a lot (https://gitlab.gnome.org/GNOME/gnome-build-meta/issues/168)

Despite of this, the Lutris flatpak is mostly ready to use.

LeandroStanger commented:
bot, build net.lutris.Lutris

LeandroStanger commented:
```
Build 3321 failed
```
https://github.com/lutris/lutris/pull/2134

gasinvein commented:
bot, build net.lutris.Lutris

bibschxz commented:
flatpak install --user https://dl.flathub.org/build-repo/3323/net.lutris.Lutris.flatpakref

link is not working for me

LeandroStanger commented:
bot, build net.lutris.Lutris

gasinvein commented:
bot, build net.lutris.Lutris

fishxz commented:
is this lutris version somehow broken? if i try to install battle.net launcher, the installation starts for 2 seconds and just quits... leaves me with uninstalled battle.net launcher.

lionirdeadman commented:
> is this lutris version somehow broken? if i try to install battle.net launcher, the installation starts for 2 seconds and just quits... leaves me with a not installed battle.net launcher.
> 
> also lutris tells me there is no wine installed on my system, im sure it didnt said that before.

This is https://github.com/RobLoach/net.lutris.Lutris/issues/15, there's a reason this is not on Flathub yet :wink:

gasinvein commented:
@lionirdeadman I hope it's not. We aren't going to support prebuilt wine, we should build it ourselves instead, see RobLoach/net.lutris.Lutris#44 for details.

TingPing commented:
> We aren't going to support prebuilt wine, we should build it ourselves instead,

What do you think Lutris is? It downloads massive prebuilt runtimes for every version of WINE, that is a core feature.

LeandroStanger commented:
bot, build net.lutris.Lutris

gasinvein commented:
@TingPing  Well, I hope the upstream to get involved and, in best scenario, target FD.o SDK for their wine builds, or at least make sure these builds don't rely on installing arbitary libs via package manager.
Otherwise, we'll end up with enormous amount of places where an ABI could break.

TingPing commented:
Yes, personally I would want to block this until upstream is involved, otherwise I agree its unsupportable.

gasinvein commented:
@TingPing Can we publish only on flathub-beta for now, and move to stable once/if the upstream gets involves? It would be helpful. Otherwise, feel free to put a "blocked" sticker on this PR.

TingPing commented:
I suppose I'm not against being in the beta branch.

gasinvein commented:
Good. Should I ask @nedrichards to make [this](https://github.com/flathub/flathub/pull/926#issuecomment-497839518) happen?

nedrichards commented:
@gasinvein I'm not against that either. @TingPing (if he has time), @barthalion and the others can do it as well if they want.

nedrichards commented:
OK, [a repo has been created](https://github.com/flathub/net.lutris.Lutris) and invitations sent - let us know if we need more. It only has a beta branch. When you're ready to go fully public then you can create a master branch and release there too.

gasinvein commented:
@nedrichards Thanks!
Can we submit ertensions for Lutris flatpak while there is beta branch only?

LeandroStanger commented:
@nedrichards Send invitation to @strycore
